packages/astro/src/vite-plugin-astro/hmr.ts

```diff
@@ -68,25 +58,39 @@ export async function handleHotUpdate(
 	}
 }
 
-function isStyleOnlyChanged(oldResult: CompileResult, newResult: CompileResult) {
-	return (
-		normalizeCode(oldResult.code) === normalizeCode(newResult.code) &&
-		// If style tags are added/removed, we need to regenerate the main Astro file
-		// so that its CSS imports are also added/removed
-		oldResult.css.length === newResult.css.length &&
-		!isArrayEqual(oldResult.css, newResult.css)
-	);
-}
+const frontmatterRE = /^\-\-\-.*?^\-\-\-/ms;
+const scriptRE = /<script(?:\s.*?)?>.*?<\/script>/gs;
+const styleRE = /<style(?:\s.*?)?>.*?<\/style>/gs;
+
+function isStyleOnlyChanged(oldCode: string, newCode: string) {
+	if (oldCode === newCode) return false;
+
+	// Before we can regex-capture style tags, we remove the frontmatter and scripts
+	// first as they could contain false-positive style tag matches. At the same time,
+	// we can also compare if they have changed and early out.
+
+	// Strip off and compare frontmatter
+	let oldFrontmatter = '';
+	let newFrontmatter = '';
+	oldCode = oldCode.replace(frontmatterRE, (m) => ((oldFrontmatter = m), ''));
+	newCode = newCode.replace(frontmatterRE, (m) => ((newFrontmatter = m), ''));
```

Comment:
If we do that, I think we'll miss out on the early checks in this function for perf though 🤔 Maybe I'm missing something for what you have in mind.

Reply:
I agree that this `replace()` + mutation pattern is quite difficult to read, but I do see how it's a useful one-liner in this case. I'm okay with it if you think this is the cleanest approach!

Reply:
> If I were to read the code in the future, it would a surprise to see that `newCode` is manipulated inside a function that returns a boolean.

This function that mutates `newCode` won't affect the value of `newCode` when used in `compile`. JavaScript passes strings to parameters as a "copy", so mutating it inside the function only stays within it. It's kinda a trick, but I found it's slightly more performant as you don't have to allocate another variable.